pkg/resource/deploy/step.go

```diff
@@ -498,6 +484,27 @@ func (s *DeleteStep) Apply() (resource.Status, StepCompleteFunc, error) {
 		}); err != nil {
 			return rst.Status, nil, err
 		}
+
+		// Delete steps may occur as part of a replacement chain (where a
+		// replacement is effectively implemented as a combination of a create and a
+		// delete). The two combinations (create first, or create last) give us the
+		// two kinds of replacements Pulumi currently supports:
+		// delete-before-replace and delete-after-replace.
+		//
+		// In the case of a delete-before-replace, we want the resource to remain in
+		// the persistence layer throughout the replace operation, even at the point
+		// where the delete has occurred but the create has not (yet). This is the
+		// purpose of `resource.State`'s `PendingReplacement` field, which we set
+		// here. Note that we don't want to set this before the provider delete
+		// call. Doing this means that we'll elide a delete next time we operate on
+		// the stack, since `PendingReplacement`'s purpose is to indicate that a
+		// delete has already been performed and that an appropriate create needs to
+		// be carried out to complete the operation.
+		if s.pendingReplacement {
+			s.old.Lock.Lock()
+			s.old.PendingReplacement = true
+			s.old.Lock.Unlock()
+		}
```

Comment:
Do we not need to do this for previews and the other cases above as well?  Since doing nothing would be a "success"?  Or does it not matter?

Reply:
Oh wow I think this might be the catch of the century. I have moved this outside the `if`. It's therefore still after the `Delete` succeeds, but should handle the other cases as they were being handled before (where it was unconditionally set before this `if`). To be clear, I _think_ the behaviour we want/have is:

* Preview -- no-op, since we are not doing anything and no state will change (?)
* External -- no-op, since we are not managing it and no state will change (?)
* Retain on delete -- we are not going to call provider delete, but state will change and thus we need to set pending replacement
* Deleted with -- we are not going to call provider delete, but state will change and thus we need to set pending replacement